substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/sampler/SamplerSampleWriter.java

```diff
@@ -181,21 +183,37 @@ private static boolean accommodate(SamplerSampleWriterData data, UnsignedWord un
             cancel(data);
             return false;
         }
-        JfrThreadLocal.setSamplerBuffer(newBuffer);
-
-        /* Copy the uncommitted content of old buffer into new one. */
-        UnmanagedMemoryUtil.copy(data.getStartPos(), SamplerBufferAccess.getDataStart(newBuffer), uncommitted);
-
-        /* Put in the stack with other unprocessed buffers and send a signal to the JFR recorder. */
         SamplerBuffer oldBuffer = data.getSamplerBuffer();
-        SubstrateJVM.getSamplerBufferPool().pushFullBuffer(oldBuffer);
-        SubstrateJVM.getRecorderThread().signal();
-
-        /* Reinitialize data structure. */
-        data.setSamplerBuffer(newBuffer);
-        reset(data);
-        increaseCurrentPos(data, uncommitted);
-        return true;
+        BufferNode oldNode = oldBuffer.getNode();
+
+        // Lock here to avoid race with flushing thread which might be processing
+        BufferNodeAccess.lockNoTransition(oldNode);
```

Comment:
Actually, I don't think that we need any `BufferNode`-based locking for `SamplerBuffer`s:
- the buffers have a fixed size and are never resized
- in-use/full buffers are only processed by threads that hold the `JfrChunkWriter` lock
- only the `SamplerBufferPool` frees buffers

So, the following should be safe:
- a flushing thread can access the data below the committed position without any locking
- `SamplerSampleWriter.accommodate(...)` may access the uncommitted data without any locking

Reply:
Ok I agree. I've removed the node-level locking for the SamplerBuffers